### PR TITLE
[Feat] Add Fault Tolerant Training for ValidationLoop

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/9130

This PR would be likely splitted in smaller pieces. First pass reviews on the logic structure.

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -98,7 +105,7 @@ def reset(self) -> None:
         # track epoch output
         self._epoch_output = [[] for _ in range(self.batch_loop.num_active_optimizers(self.total_batch_idx))]
 
-        if not self.restarting or self._num_training_batches_reached():
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:101]: I ran your branch and I see the same issue here again #8835

File: pytorch_lightning/trainer/progress.py [hunk 2]
```diff
@@ -153,6 +153,10 @@ def load_state_dict(self, state_dict: dict) -> None:
         self.total.load_state_dict(state_dict["total"])
         self.current.load_state_dict(state_dict["current"])
 
+    def reset_on_restart(self):
+        self.total.reset_on_restart()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:157]: what does it mean to reset the total counter? I don't understand this

File: pytorch_lightning/loops/optimization/closure.py [hunk 3]
```diff
@@ -188,3 +189,25 @@ def closure(self, *args: Any, **kwargs: Any) -> ClosureResult:
...
+    def consume_result(self) -> ClosureResult:
+        return self._closure.consume_result()
+
+    def __call__(self):
+        # enable fault tolerant during closure execution
+        with self._trainer._fault_tolerant_supported(enable=True):
+            loss = self._closure()
+        self._fault_tolerant_possible = self._trainer._fault_tolerant_possible
+        # prevent fault tolerant during parameters update.
+        self._trainer._fault_tolerant_possible = False
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimization/closure.py]: do we expect a failure during the parameter update in the optimizer?

Comment (awaelchli) [reply pytorch_lightning/loops/optimization/closure.py]: don't we want the lightning optimizer to take care of this?

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -464,18 +464,10 @@ def _capture_metadata_collate(samples: List, dataset: Dataset, default_collate:
...
-
-    elif isinstance(dataset, CaptureMapDataset):
-        samples, states = zip(*samples)
-        data = default_collate(samples)
-        metadata = states[-1]
-    else:
-        return default_collate(samples)
-
-    return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
+    data = default_collate(samples)
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:467]: I already made these changes based on your recommendation in a separate PR: #9484 
this is not relevant to this PR as far as I can tell.

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 5]
```diff
@@ -278,6 +279,8 @@ def _run_optimization(
         # ------------------------------
         # gradient update with accumulated gradients
         else:
+            if _fault_tolerant_training():
+                closure = _FaultToleranceClosureExecutor(closure, self.trainer)
```
Comment (carmocca) [inline pytorch_lightning/loops/optimization/optimizer_loop.py:283]: Can you explain this? Is it to avoid saving on exception if `optimizer.step` hasn't been called?

Why do we need it?
How is it related to the validation loop?

File: pytorch_lightning/loops/fit_loop.py [hunk 6]
```diff
@@ -263,8 +263,9 @@ def teardown(self) -> None:
 
     def on_save_checkpoint(self) -> Dict:
         state_dict = super().on_save_checkpoint()
-        # TODO: update has_completed to its proper value
-        state_dict["dataloader_state_dict"] = self.trainer.train_dataloader.state_dict(has_completed=False)
+        # FIXME(@tchaton) Should pass has_completed=True when iterator is exhausted ?
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:266]: I think @awaelchli recently changed this

**Review by awaelchli — COMMENTED** [review]
are we going to partition every loop into 
`fault_tolerant_supported(true)`
and
`fault_tolerant_supported(false)`
? 
I don't have the full picture on this.

File: pytorch_lightning/core/optimizer.py [hunk 7]
```diff
@@ -121,6 +121,16 @@ def toggle_model(self, sync_grad: bool = True):
             yield
             self._untoggle_model()
 
+    @contextmanager
+    def __fault_tolerant_supported(self) -> Generator:
+        """The closure should prevent execution fault tolerance within the parameters update.
+
+        Restore the fault tolerant state accordingly
```
Comment (carmocca) [inline pytorch_lightning/core/optimizer.py:128]: Is this asuming that the inner code (the one that yields) will change the value of `self._trainer._fault_tolerant_possible`?

Is the purpose here to set the previous value in case that happens?

This is very ugly in terms of side effects and misdirection.

File: pytorch_lightning/loops/base.py [hunk 8]
```diff
@@ -217,7 +217,7 @@ def _load_from_state_dict(
             if isinstance(v, BaseProgress):
                 v.load_state_dict(state_dict[key])
                 if restart_progress:
-                    apply_to_collection(v, Progress, lambda p: p.current.reset_on_restart())
+                    apply_to_collection(v, BaseProgress, lambda p: p.reset_on_restart())
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:220]: Why is this necessary?

Not reseting `p.total` is done on purpose.

Why do we need this `restart_progress` flag and this code here instead of relying on

```python
def reset():
    if self.restarting:
         reset_something()
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 9]
```diff
@@ -76,7 +76,14 @@ def done(self) -> bool:
         signals to stop (e.g. by early stopping).
         """
         max_steps_reached = self.max_steps is not None and self.global_step >= self.max_steps
-        return max_steps_reached or self.trainer.should_stop or self._num_training_batches_reached()
+        # ßdone should be True only when `batch_loop` and `val_loop` are done.
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:79]:
```suggestion
        # done should be True only when `batch_loop` and `val_loop` are done.
```

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 10]
```diff
@@ -76,7 +76,14 @@ def done(self) -> bool:
...
+        should_check_val = self.restarting and self._should_check_val_fx(self.batch_idx, self.is_last_batch)
+        return (
+            max_steps_reached
+            or self.trainer.should_stop
+            or self._num_training_batches_reached()
+            and not should_check_val
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:85]: why is adding this necessary?

running validation should have happened if it had to in the previous iteration

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:101]: bad merge? this was recently added by @awaelchli

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:101]: Interesting. This might be problematic as it would reset the progress even if validation hasn't been run. I need to find about the right solution there.

File: pytorch_lightning/trainer/progress.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:157]: Good catch ! This is a bug.

File: pytorch_lightning/loops/optimization/closure.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/loops/optimization/closure.py]: The LightningOptimizer can't handle this logic as the closure is being passed the accelerators. The only solution I found was to wrap the closure on optimizer.step call.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:101]: we could save to the state_dict if validation has run or not

File: pytorch_lightning/loops/fit_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/loops/fit_loop.py:266]: Reverted back.

File: pytorch_lightning/loops/base.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:220]: `OptimizationProgress` was not properly resetted. It took me some to find this. I think it is safer to always call the BaseProgress and rely on the reset_on_restart to not reset progress.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:85]: `done` shouldn't be `True` if validation has run yet in advance_end. Therefore, `should_check_val` needs to be added to enable restarting within the validation loop.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:101]: No, it wasn't a bad merge. Without this, restart on validation loop won't work.
But I would have to find a way to make both work together.

File: pytorch_lightning/loops/optimization/optimizer_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:283]: This is to avoid saving a checkpoint on failure if the weights were being updated.

File: pytorch_lightning/core/optimizer.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:128]: Yes, this was a solution to complete the `self._trainer._fault_tolerant_possible=False` applied after the closure is being run.

**Comment by tchaton** [comment]
Hey @awaelchli @carmocca,

From your comments, you aren't too fan of the `fault_tolerant_supported`, therefore I will drop it for now.

And I will try to find a way to make @awaelchli fix works with restart in validation loop.

Best,
T.C

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:101]: Yes, this is what I ended up doing.

File: pytorch_lightning/loops/base.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:220]: IMO, I prefer the loop to take care of sanitising progress.
